src/Server/HTTPHandler.cpp

```diff
@@ -596,8 +596,15 @@ void HTTPHandler::processQuery(
     customizeContext(request, context);
 
     executeQuery(*in, *used_output.out_maybe_delayed_and_compressed, /* allow_into_outfile = */ false, context,
-        [&response] (const String & current_query_id, const String & content_type, const String & format, const String & timezone)
+        [&session, &response] (const String & current_query_id, const String & content_type, const String & format, const String & timezone)
         {
+            /// Release the session when the query is done.
```

Comment:
This callback [is called](https://github.com/ClickHouse/ClickHouse/blob/ddf0fe30b37ff989a2187f53bbc34e783551c9c2/src/Interpreters/executeQuery.cpp#L850) before query pipeline [execution](https://github.com/ClickHouse/ClickHouse/blob/ddf0fe30b37ff989a2187f53bbc34e783551c9c2/src/Interpreters/executeQuery.cpp#L861), so query is not done. I'm not sure if it's safe to release a session here.